### PR TITLE
Add parentRelationType to AnalysisContext

### PR DESCRIPTION
**Pull request by kokosing**
Add parentRelationType to AnalysisContext

Thanks to that ExpressionAnalyzer is able to recognize that reference to
outer relation was used.

### CODE REVIEW

**Comment by kokosing** [comment]
@martint Can you please take a look

**Comment by cberner** [comment]
Can you check Travis? It looks like there are checkstyle failures

**Comment by kokosing** [comment]
Sorry about that. Now it should be ok.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -307,7 +307,12 @@ protected Type visitDereferenceExpression(DereferenceExpression node, StackableA
                     return field.getType();
                 }
 
-                assertColumnPrefix(qualifiedName, node);
+                if (!isColumnPrefix(qualifiedName, tupleDescriptor)) {
+                    if (isReferenceToOuterRelation(context.getContext(), qualifiedName)) {
+                        throw new SemanticException(NOT_SUPPORTED, node, "Accessing outer relation symbol '%s', while correlated subqueries are not yet supported", node);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: The notion of "symbol" is an internal detail of the analyzer/planner. It'd be better to say "column" or "field". I would rephrase this as:

"Correlated subqueries are not yet supported. Invalid column reference: %s"

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 2]
```diff
@@ -332,19 +337,37 @@ protected Type visitDereferenceExpression(DereferenceExpression node, StackableA
...
-                    // The AMBIGUOUS_ATTRIBUTE exception will be thrown later with the right node if matches.size() > 1
-                    return;
+                    return true;
                 }
             }
+            return false;
+        }
 
-            throw createMissingAttributeException(node);
+        private boolean isReferenceToOuterRelation(AnalysisContext context, QualifiedName qualifiedName)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: This can be simplified slightly to:

```java
AnalysisContext current = context;
while (current != null) {
    RelationType type = current.getParentRelationType();
    if (!type.resolveFields(qualifiedName).isEmpty()) {
        return true;
    }
    if (isColumnPrefix(qualifiedName, type)) {
        return true;
    }
    current = current.getParent().orElse(null);
}
return false;
```

**Comment by martint** [comment]
Applied those changed and merged. Thanks!

**Comment by kokosing** [comment]
Nice. Thanks.

sob., 30.04.2016, 18:02 użytkownik Martin Traverso <notifications@github.com>
napisał:
